### PR TITLE
A new service container architect

### PR DESCRIPTION
**Pull request by zackliu**
* A new architect around service connection / factory / container
* Add default connection and on-demand connection support
* Add route target support

This PR doesn't change the actual way to reconnect because it will change lots of UT. To simplify the PR, keep reconnect logic as-is.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/StrongServiceConnectionContainer.cs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+            }
+
+            int index = _serviceConnections.IndexOf(connection);
+            if (index != -1)
+            {
+                lock (_lock)
+                {
+                    if (_onDemandServiceConnections.Count != 0)
+                    {
+                        _serviceConnections[index] = _onDemandServiceConnections[0];
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/StrongServiceConnectionContainer.cs]: should check is connected.

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ConnectionFactory.cs [hunk 2]
```diff
@@ -54,10 +54,14 @@ public async Task<ConnectionContext> ConnectAsync(TransferFormat transferFormat,
...
         }
 
-        private Uri GetServiceUrl(string connectionId)
+        private Uri GetServiceUrl(string connectionId, string target)
         {
             var baseUri = new UriBuilder(_provider.GetServerEndpoint(_hubName));
             var query = "cid=" + connectionId;
+            if (target != null)
+            {
+                query = query + $"&target={target}";
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ConnectionFactory.cs]: url encode (data string)

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.SignalR.Common;
+using Microsoft.Azure.SignalR.Protocol;
+
+namespace Microsoft.Azure.SignalR
+{
+    abstract class ServiceConnectionContainerBase : IServiceConnectionContainer, IServiceConnectionManager
+    {
+        private static readonly int MaxReconnectBackOffInternalInMilliseconds = 1000;
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs]: remove

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        private static TimeSpan ReconnectInterval =>
+            TimeSpan.FromMilliseconds(StaticRandom.Next(MaxReconnectBackOffInternalInMilliseconds));
+
+        protected readonly IServiceConnectionFactory ServiceConnectionFactory;
+        protected readonly IConnectionFactory ConnectionFactory;
+        protected readonly List<IServiceConnection> ServiceConnections;
+        protected readonly int Count;
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs]: naming

File: src/Microsoft.Azure.SignalR.Common/Interfaces/IServiceConnectionManager.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.SignalR
+{
+    interface IServiceConnectionManager
+    {
+        IEnumerable<IServiceConnection> CreateServiceConnection(int count = 1);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Interfaces/IServiceConnectionManager.cs]: >CreateServiceConnection [](start = 40, length = 23)

naming: `Connections` for `IEnumerable`

File: src/Microsoft.Azure.SignalR.Common/Interfaces/IServiceConnectionFactory.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.SignalR
+{
+    interface IServiceConnectionFactory
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Interfaces/IServiceConnectionFactory.cs]: >IServiceConnectionFactory [](start = 14, length = 25)

naming: this is `Creator` the other one as `Factory`?

File: src/Microsoft.Azure.SignalR.Common/Interfaces/IServiceConnectionManager.cs (hunk 5 above)
Comment (zackliu) [reply src/Microsoft.Azure.SignalR.Common/Interfaces/IServiceConnectionManager.cs]: We decide to only support creating one connection

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ConnectionFactory.cs (hunk 2 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/ConnectionFactory.cs]: $"{query}&target={target}"

---
In reply to: [248543449](https://github.com/Azure/azure-signalr/pull/342#discussion_r248543449) [](ancestors = 248543449)

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs [hunk 7]
```diff
@@ -181,15 +186,37 @@ protected Task OnServiceErrorAsync(ServiceErrorMessage serviceErrorMessage)
...
-        private async Task<bool> StartAsyncCore()
+        protected Task OnPingMessageAsync(PingMessage pingMessage)
+        {
+            if (pingMessage.Messages.Length == 0)
+            {
+                return Task.CompletedTask;
+            }
+
+            int index = 0;
+            while (index < pingMessage.Messages.Length)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: >index [](start = 19, length = 5)

`< length -1` for safety?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs [hunk 8]
```diff
@@ -181,15 +186,37 @@ protected Task OnServiceErrorAsync(ServiceErrorMessage serviceErrorMessage)
...
+            {
+                if (pingMessage.Messages[index] == PingTargetKey &&
+                    !string.IsNullOrEmpty(pingMessage.Messages[index + 1]))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: >pingMessage.Messages[index + 1] [](start = 42, length = 31)

target not used?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs [hunk 9]
```diff
@@ -181,15 +186,37 @@ protected Task OnServiceErrorAsync(ServiceErrorMessage serviceErrorMessage)
...
+                {
+                    var connection = _serviceConnectionManager.CreateServiceConnection();
+                    return connection.First().StartAsync();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: >.First() [](start = 37, length = 8)

why first?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs [hunk 10]
```diff
@@ -70,15 +74,16 @@ public ServiceConnectionBase(IServiceProtocol serviceProtocol, ILogger logger, s
 
             _cachedPingBytes = serviceProtocol.GetMessageBytes(PingMessage.Instance);
             _handshakeRequest = new HandshakeRequestMessage(serviceProtocol.Version, (int)connectionType);
+            _serviceConnectionManager = serviceConnectionManager;
```
[the remark this answers is not in the document]
Comment (zackliu) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: For tests in this PR, let the null be valid

File: src/Microsoft.Azure.SignalR.AspNet/OwinExtensions.cs [hunk 11]
```diff
@@ -242,7 +242,7 @@ private static ServiceHubDispatcher PrepareAndGetDispatcher(HubConfiguration con
             configuration.Resolver.Register(typeof(IProtectedData), () => provider);
 
             var scm = new ServiceConnectionManager(applicationName, hubs);
-            configuration.Resolver.Register(typeof(IServiceConnectionManager), () => scm);
+            configuration.Resolver.Register(typeof(Microsoft.Azure.SignalR.AspNet.IServiceConnectionManager), () => scm);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/OwinExtensions.cs]: >IServiceConnectionManager [](start = 82, length = 25)

rename to `IHubConnectionManager`?

Comment (zackliu) [reply src/Microsoft.Azure.SignalR.AspNet/OwinExtensions.cs]: Rename it in a seperate PR

**Pull request merged**